state/protocol/protocol_state/kvstore/models.go

```diff
@@ -0,0 +1,83 @@
+
+// modelv0 is v0 of the Protocol State key-value store.
+// This model version is not intended to ever be the latest version supported by
+// any software version. Since it is important that the store support managing
+// different model version, this is here so that we can test the implementation
+// with multiple supported KV model versions from the beginning.
+type modelv0 struct{}
+
+var _ Reader = new(modelv0)
+var _ API = new(modelv0)
+
+// VersionedEncode encodes the key-value store, returning the version separately
+// from the encoded bytes.
+// No errors are expected during normal operation.
+func (model *modelv0) VersionedEncode() (uint64, []byte, error) {
+	return versionedEncode(model.GetProtocolStateVersion(), model)
+}
+
+// GetProtocolStateVersion returns the Protocol State version.
+func (model *modelv0) GetProtocolStateVersion() uint64 {
+	return 0
+}
+
+// GetInvalidEpochTransitionAttempted returns ErrKeyNotSupported.
+func (model *modelv0) GetInvalidEpochTransitionAttempted() (bool, error) {
+	return false, ErrKeyNotSupported
+}
+
+// SetInvalidEpochTransitionAttempted returns ErrKeyNotSupported.
+func (model *modelv0) SetInvalidEpochTransitionAttempted(_ bool) error {
+	return ErrKeyNotSupported
+}
+
+// modelv1 is v1 of the Protocol State key-value store.
+// This represents the first model version which will be considered "latest" by any
+// deployed software version.
+type modelv1 struct {
+	// InvalidEpochTransitionAttempted encodes whether an invalid epoch transition
+	// has been detected in this fork. Under normal operations, this value is false.
+	// Node-internally, the EpochFallback notification is emitted when a block is
+	// finalized that changes this flag from false to true.
+	//
+	// Currently, the only possible state transition is false → true.
+	InvalidEpochTransitionAttempted bool
+}
+
+var _ Reader = new(modelv1)
+var _ API = new(modelv1)
+
+// VersionedEncode encodes the key-value store, returning the version separately
+// from the encoded bytes.
+// No errors are expected during normal operation.
+func (model *modelv1) VersionedEncode() (uint64, []byte, error) {
+	return versionedEncode(model.GetProtocolStateVersion(), model)
+}
+
+// GetProtocolStateVersion returns the Protocol State version.
+func (model *modelv1) GetProtocolStateVersion() uint64 {
+	return 1
+}
+
+// GetInvalidEpochTransitionAttempted returns the InvalidEpochTransitionAttempted flag.
+// This key must have a value set and will never return ErrKeyNotSet.
+// No errors are expected during normal operation.
+func (model *modelv1) GetInvalidEpochTransitionAttempted() (bool, error) {
+	return model.InvalidEpochTransitionAttempted, nil
+}
+
+// SetInvalidEpochTransitionAttempted sets the InvalidEpochTransitionAttempted flag.
+// No errors are expected during normal operation.
+func (model *modelv1) SetInvalidEpochTransitionAttempted(attempted bool) error {
+	model.InvalidEpochTransitionAttempted = attempted
+	return nil
+}
```

Comment:
My comment from earlier put slightly differently: 
* why not have an interface (let's call it `EpochRecovery`) that `modelv1` implements but `modelv0` doesn't? 

Not sure. I guess it is something we can change later too - as part of a rolling software upgrade 😅.

Reply:
If I understand correctly, there are two aspects to your two comments:
1. Why does `modelv0` implement setter/getter functions for keys that don't exist in it?
2. Why is the `InvalidEpochTransitionAttempted` state in the kv store, rather than in the existing protocol state machine? 

Regarding 1, all supported models implementing the latest public interface is an important part of how we plan to support multiple versions. The idea is:
- components outside of the `kvstore` package use instances of `kvstore.API` and `kvstore.Reader` -- they don't have versioned interfaces to the key-value store
- within the `kvstore` package we operate on the versioned concrete types and handle encoding/decoding in a versioned way

Suppose our software supports versions 1 and 2, and the current `ProtocolStateVersion` is 1. We are operating on instances of `modelv1`, and storing this model to disk within `kvstore`, but the public `API` interface includes keys from `modelv2` as well. In order for instances of `modelv1` to be used as an `API` instance externally, we implement the `v2` methods, but return sentinels. 

Regarding 2, I have added `InvalidEpochTransitionAttempted` into the kv store as an example, using a piece of state that already exists. Once we integrate the kvstore and protocol state machines, I think it will be clearer whether it makes sense to keep it here or move it back to the protocol state machine.